src/d3/xy-chart.d3.ts

```diff
@@ -66,31 +104,88 @@ export abstract class XYChartD3 extends BaseD3<RenderOptions> {
 
     this.xAxisG = this.svg
       .append('g')
-      .attr('transform', `translate(0,${height - marginBottom})`);
+      .attr('transform', `translate(0,${bottom})`);
     this.yAxisG = this.svg
       .append('g')
-      .attr('transform', `translate(${marginLeft},0)`);
+      .attr('transform', `translate(${left},0)`);
   }
 
-  protected updateAxis(data: Datum[]) {
-    this.scaleX.domain(d3.extent<Datum, Date>(data, d => d.date) as [Date, Date]);
-    this.scaleY.domain([0, d3.max(data, d => d.value)!]);
+  protected updateAxis(data: ResultRow[]) {
+    const { fontSizeSmall } = XYChartD3;
+
+    this.scaleX.domain(d3.extent<ResultRow, Date>(data, d => d.categories.date) as [Date, Date]);
+    this.scaleY.domain([0, d3.max(data, d => Math.max(...Object.values(d.values)))!]);
 
     this.xAxisG
       .transition(this.transition)
       .call(this.xAxis)
-      .attr('font-size', 12);
+      .attr('font-size', fontSizeSmall);
     this.yAxisG
       .transition(this.transition)
       .call(this.yAxis)
-      .attr('font-size', 12);
+      .attr('font-size', fontSizeSmall);
+  }
+
+  protected renderLegend() {
+    const { height } = this.renderOptions;
+    const { legendHeight, legendPaddingTop, padding } = XYChartD3;
+
+    const top = height - padding - legendHeight + legendPaddingTop;
+    const left = padding;
+
+    this.legendG = this.svg
+      .append('g')
+      .attr('transform', `translate(${left},${top})`);
+  }
+
+  protected updateLegend(legendItems: LegendItem[]) {
+    const { legendIconWidth, legendHeight, fontSizeMedium } = XYChartD3;
+
+    this.legendG.html('');
+
+    let offsetX = 0;
+    for (const { label, style } of legendItems) {
+      const itemG = this.legendG
+        .append('g')
+        .attr('transform', `translate(${offsetX}, 0)`);
+      itemG
```

Comment:
nit: continue chained operation?

Reply:
I don't think they should be chained. Line 152 is appending a new `line` element to `itemG`.